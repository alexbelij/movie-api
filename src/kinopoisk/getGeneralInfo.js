/* @flow */

import R from 'ramda';

import { normalizeCdnImageUrl } from './utils';
import connector from './connector';

type KinopoiskApi$GetFilmResponse = {
  filmID?: ?string,
  imdbID?: ?string,
  nameRU?: ?string,
  nameEN?: ?string,
  posterURL?: ?string,
  year?: ?string,
  description?: ?string,
  filmLength?: ?string,
  genre?: ?string,
  ratingAgeLimits?: ?string,
  ratingMPAA?: ?string,
  ratingData: {
    rating?: ?string,
    ratingVoteCount?: ?string,
    ratingIMDb?: ?string,
    ratingIMDbVoteCount?: ?string,
    ratingFilmCritics?: ?string,
    ratingFilmCriticsVoteCount?: ?string,
  },
};

const parseRuntime = (rawRuntime: ?string): number => (
  rawRuntime
    ? R.pipe(
        R.split(':'),
        R.slice(0, 2),
        ([hours, minutes]: [string, string]): number => R.sum([
          parseInt(hours, 10) * 60,
          parseInt(minutes, 10),
        ]),
        (runtime: number) => (runtime === 0 ? NaN : runtime),
      )(rawRuntime)
    : NaN
);

const getGeneralInfo = async (id: number) => {
  const res: ?KinopoiskApi$GetFilmResponse =
    await connector.get('getFilm', { filmID: id });

  if (!res) return null;

  const {
    filmID,
    imdbID,
    nameRU,
    nameEN,
    posterURL,
    year,
    description,
    filmLength,
    genre,
    ratingAgeLimits,
    ratingMPAA,
    ratingData = {},
  } = res;

  return {
    kpId: parseInt(filmID, 10),
    imdbID,
    title: {
      ru: nameRU,
      original: nameEN,
    },
    posterUrl: posterURL ? normalizeCdnImageUrl(posterURL) : null,
    year: parseInt(year, 10),
    synopsis: {
      ru: description,
    },
    runtime: parseRuntime(filmLength),
    genres: (genre || '').split(', '),
    ageRating: parseInt(ratingAgeLimits, 10),
    mpaaRating: ratingMPAA,
    kpRating: parseFloat(ratingData.rating),
    kpRatingVoteCount: parseInt(ratingData.ratingVoteCount, 10),
    imdbRating: parseFloat(ratingData.ratingIMDb),
    imdbRatingVoteCount: parseInt(ratingData.ratingIMDbVoteCount, 10),
    rtCriticsRating: parseInt(ratingData.ratingFilmCritics, 10),
    rtCriticsRatingVoteCount: parseInt(
      ratingData.ratingFilmCriticsVoteCount,
      10,
    ),
  };
};

export { parseRuntime as __parseRuntime };
export default getGeneralInfo;
