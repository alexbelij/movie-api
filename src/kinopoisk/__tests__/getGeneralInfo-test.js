/* @flow */

import R from 'ramda';

import getGeneralInfo from '../getGeneralInfo';

const kpId = 714888; // Star Wars: The Force Awakens

const childTypesForArray = R.memoize(R.pipe(R.map(R.type), R.uniq));
const modelFromObject = (obj: ?Object) => R.map(
  R.cond([
    [
      R.allPass([
        R.pipe(R.type, R.equals('Array')),
        R.pipe(childTypesForArray, R.length, R.equals(1)),
      ]),
      (val: Array<any>) => `Array<${R.pipe(childTypesForArray, R.head)(val)}>`,
    ],
    [R.pipe(R.type, R.contains(R.__, ['Object', 'Array'])), modelFromObject],
    [R.T, R.type],
  ]),
  obj,
);

describe('kinopoisk/getGeneralInfo', () => {
  it('fetches general movie info from kinopoisk for a given id', async () => {
    expect(modelFromObject(await getGeneralInfo(kpId))).toMatchSnapshot();
  });
});
