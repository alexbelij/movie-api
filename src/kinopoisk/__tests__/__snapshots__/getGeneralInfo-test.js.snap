exports[`kinopoisk/getGeneralInfo fetches general movie info from kinopoisk for a given id 1`] = `
Object {
  "ageRating": "Number",
  "genres": "Array<String>",
  "imdbID": "String",
  "imdbRating": "Number",
  "imdbRatingVoteCount": "Number",
  "kpId": "Number",
  "kpRating": "Number",
  "kpRatingVoteCount": "Number",
  "mpaaRating": "String",
  "posterUrl": "String",
  "rtCriticsRating": "Number",
  "rtCriticsRatingVoteCount": "Number",
  "runtime": "Number",
  "synopsis": Object {
    "ru": "String",
  },
  "title": Object {
    "original": "String",
    "ru": "String",
  },
  "year": "Number",
}
`;
